### ISSUE TITLE
Remap of magit-submodule suffix in magit-dispatch transient prefix causes forge to print an error

### ISSUE THREAD

**Issue by sin-ack** [emacs-evil/evil-collection#543]
Hi. This line:
https://github.com/emacs-evil/evil-collection/blob/6709c1ec4118c8721df43ea6708ae45ebbc01fd3/modes/magit/evil-collection-magit.el#L574
remaps `magit-submodule`'s key in the `magit-dispatch` transient prefix from `o` to `'`. Since magit/forge@9fbd00e, forge will now print an error when a Magit buffer is first visited:
```
Cannot insert ("N" "Forge" forge-dispatch) into magit-dispatch; o not found
```
as it cannot find the `o` key to insert its own key before. Not quite sure how to approach fixing this, because Forge can technically do that just fine, but remapping the keys should be fine too.

**Comment by jojojames** [comment]
Hmnn.. this one is tricky since it seems like those changes are at the top level in forge.el. Any suggestions @sin-ack ?

**Comment by tshu-w** [comment]
This only happens if we run `(evil-collection-magit-setup)` before `(require 'forge)`, we should `(require 'forge)` before `(evil-collection-magit-init)`:
```emacs-lisp
  (defun evil-collection-magit-setup@override ()
    "Set up `evil' bindings for `magit'."

    (evil-collection-define-key 'normal 'magit-blame-mode-map
      "q" 'magit-blame-quit)
    (evil-collection-define-key 'normal 'magit-blame-read-only-mode-map
      "q" 'magit-blame-quit)

    (require 'forge)
    (evil-collection-magit-init))
```
or seperate forge setup from magit in evil-collections.

**Comment by jojojames** [comment]
We don't want to forcer equire forge as not everyone uses it (me included).

**Comment by jojojames** [comment]
Thanks @tarsius ! I made some initial changes here (https://github.com/emacs-evil/evil-collection/pull/549/files) and it looks okay to me but I don't really use forge so not sure if anything is missing. @sin-ack Can you pull @tarsius 's changes for forge on his no-defaults branch and evil-collection's forge branch and see if it works for you?

If not we can probably go with the current set of changes on both branches.

**Comment by tshu-w** [comment]
> Can you pull @tarsius 's changes for forge on his no-defaults branch and evil-collection's forge branch and see if it works for you?

This works for me.

**Comment by tarsius** [comment]
I have merged the new variables into the master branch.

I have checked out `evil-collection`'s `forge` branch and do not see any mention of these new variables, so I don't know what @tshu-w is referring to when they say that it works.

Looking at [Summary: Sequence of Actions at Startup](https://www.gnu.org/software/emacs/manual/html_node/elisp/Startup-Summary.html) we can see that in step 7 `package-activate-all` is called. Looking at `package.el` we can see that that involves loading the autoload files for the packages it installed, which is relevant because some of forges bindings involve autoloaded code:

```elisp
;;;###autoload
(defvar forge-add-default-bindings t
  "Whether to add Forge's bindings to various Magit keymaps.
If you want to disable this, then you must set this to nil before
`forge' is loaded.")

;;;###autoload
(with-eval-after-load 'magit-mode
  (when forge-add-default-bindings
    (define-key magit-mode-map "'" 'forge-dispatch)
    (define-key magit-mode-map "N" 'forge-dispatch)))
```

Then in step 14 `init.el` is loaded. It now all depends on the order in that file (resp. other files loaded by it).

Good:

```elisp
(setq forge-add-default-bindings t)
(require 'magit) ; or something to the effect
```

Bad:

```elisp
(require 'magit) ; or something to the effect
(setq forge-add-default-bindings t)
```

If you want to be really sure, then you would have to set these variables in `early-init.el`. Otherwise you depend on the user *not* doing something to the effect of:

```elisp
(require 'magit)
(require 'evil-collection)
```

**Comment by jojojames** [comment]
I did something like this to "repro" the issue.

```
+(use-package forge
+  :straight t
+  :init
+  (setq forge-add-default-bindings nil))
```

I'm assuming @tshu-w did something similar?

**Comment by tshu-w** [comment]
Yes, I mean I checkout both modified branch (now forge default) and `(setq forge-add-default-bindings nil)` before both magit and forge load, and I won't see the error message.

**Comment by jojojames** [comment]
Sounds good, closing.

**Comment by tarsius** [comment]
@jojojames I think you should add this to `evil-collection-forge.el` because users otherwise have to somehow learn that they have to add it to their init file:

```elisp
(setq forge-add-default-sections nil)
(setq forge-add-default-bindings nil)
```

Of course once you have done that its still possible that users effectively do this in their init file:

```elisp
(require 'magit)
(require 'evil-collection)
```

Which leads to a variant of the issue I mentioned above. Just something to keep in mind for when users report that bindings aren't what they expected.

**Comment by jojojames** [comment]
I'm thinking we need to document that those setq need to be set but evil-collection-forge loads after forge so those bindings have to be outside of evil-collection/before forge loads.

**Comment by tarsius** [comment]
That doesn't seem ideal to me. Maybe you could set them in another file, such as `evil-collection-magit.el`. That should be loaded before `forge` because that library requires `magit` early on.

**Comment by jojojames** [comment]
Yeah I'm not sure how to get it just right since magit is deferred on my end.

E.g.

```
(use-package magit
  :commands (magit-toplevel
             magit-status
             magit-blame-addition
             magit-log-other
             magit-find-file
             magit-find-file-other-window))
```

What would be the right way to defer forge also?

**Comment by tarsius** [comment]
E.g.

```elisp
(use-package forge
  :defer t
  :init
  (setq forge-add-default-sections nil)
  (setq forge-add-default-bindings nil))
```

But again, I think it is better if not every user has to figure that out and you can save them from having to do that like so:

```diff
diff --git a/modes/magit/evil-collection-magit.el b/modes/magit/evil-collection-magit.el
index 3f788b6..0a9d845 100644
--- a/modes/magit/evil-collection-magit.el
+++ b/modes/magit/evil-collection-magit.el
@@ -692,6 +692,14 @@ (defun evil-collection-magit-setup ()
 
   (evil-collection-magit-init))
 
+;; This has to be done before `forge' is loaded so do it here.
+(defvar forge-add-default-sections)
+(defvar forge-add-default-bindings)
+;;;###autoload
+(setq forge-add-default-sections nil)
+;;;###autoload
+(setq forge-add-default-bindings nil)
+
 ;;; evil-collection-magit.el ends soon
 (provide 'evil-collection-magit)
 ;; Local Variables:
```

**Comment by jojojames** [comment]
If I try to defer forge & magit, e.g.

```
(use-package forge
  :defer t
  :config
  (message "forge loaded"))

(use-package magit
  :commands (magit-toplevel
             magit-status
             magit-blame-addition
             magit-log-other
             magit-find-file
             magit-find-file-other-window)
  :config
  (message "magit loaded"))
```

Forge never ends up loading. (Sorry if this is getting into the debugging/troubleshooting level.)

This block looks reasonable to me though:

```
+;; This has to be done before `forge' is loaded so do it here.
+(defvar forge-add-default-sections)
+(defvar forge-add-default-bindings)
+;;;###autoload
+(setq forge-add-default-sections nil)
+;;;###autoload
+(setq forge-add-default-bindings nil)```

**Comment by tarsius** [comment]
> Forge never ends up loading.

Right. Use `:after magit` instead of `:defer t`.

> This block looks reasonable to me though:

Great :grin:

**Comment by jojojames** [comment]
I've been pretty slow on the draw, had some more time to play with this, unfortunately it looks like the loading is somewhat unexpected.

```
Turning on magit-auto-revert-mode...done
Configuring package magit...
USE-PACKAGE MAGIT :CONFIG
Configuring package magit...done
USE-PACKAGE FORGE :INIT
Loading package forge...
ADDED DEFAULT SECIONS
ADDED DEFAULT BINDINGS
EVIL-COLLECTION-FORGE SETUP
Configuring package forge...
USE-PACKAGE FORGE :CONFIG
Configuring package forge...done
Loading package forge...done (0.558s)
evil-collection magit autoload
EVIL-COLLECTION-MAGIT SETUP
Quit
```

Might be hard to grok without the print statements I added but the "ADDED DEFAULT BINDINGS" was added to forge.el where forge is binding the default keys.

Looks like magit loads, then forge loads, then evil-collection-forge then evil-collection-magit finally loads and sets the autoloads.

**Comment by tarsius** [comment]
Given a `forge.el` modified to contain:

``` emacs-lisp
;;;###autoload(message "== autoloading (defvar forge-add-default-bindings...")
(message "== proper loading (defvar forge-add-default-bindings...")
;;;###autoload
(defvar forge-add-default-bindings
  (progn (message "== setting unbound forge-add-default-bindings to t")
         t)
  "Whether to add Forge's bindings to various Magit keymaps.
If you want to disable this, then you must set this to nil before
`forge' is loaded.  You just do so even before `magit' is loaded or
else two bindings are added to `magit-mode-map'.")

(message "-- while loading forge.el: forge-add-default-bindings is %s"
	 forge-add-default-bindings)

;;;###autoload
(with-eval-after-load 'magit-mode
  (message "-- while evaling autoloaded (with-eval-after-load 'magit-mode")
  (message "-- forge-add-default-bindings is %s" forge-add-default-bindings)
  (when forge-add-default-bindings
    (message "slightly BAD add default bindings to magit-mode-map")
    (define-key magit-mode-map "'" 'forge-dispatch)
    (define-key magit-mode-map "N" 'forge-dispatch)))

(when forge-add-default-bindings
  (message "BAD add default bindings to elsewhere")
  ...)

```

, making sure `forge-autoloads.el` is updated, and this `init.el`:

``` emacs-lisp
(message "Loading %s..." load-file-name)

(setq use-package-minimum-reported-time 0.000001)
(setq use-package-verbose t)

(use-package forge
  :after magit)

(use-package magit
  :init
  (message "++ in init.el in (use-package magit ...)")
  (message "++ setting forge-add-default-bindings to nil")
  (setq forge-add-default-sections nil)
  (setq forge-add-default-bindings nil))

(message "Loading %s...done" load-file-name)

```

I get:

```
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/early-init.el...done
== autoloading (defvar forge-add-default-bindings...
== setting unbound forge-add-default-bindings to t
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/init.el...
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/custom.el (source)...done
++ in init.el in (use-package magit ...)
++ setting forge-add-default-bindings to nil
Loading package magit...
-- while evaling autoloaded (with-eval-after-load ’magit-mode
-- forge-add-default-bindings is nil
Loading package forge...
== proper loading (defvar forge-add-default-bindings...
-- while loading forge.el: forge-add-default-bindings is nil
-- while evaling autoloaded (with-eval-after-load ’magit-mode
-- forge-add-default-bindings is nil
Loading package forge...done (0.490s)
Loading package magit...done (1.748s)
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/init.el...done
Turning on magit-auto-revert-mode...done
```

Since none of the `BAD` messages occurred this demonstrates that it is possible to prevent the default bindings from being added.

If I change `init.el` to:

``` emacs-lisp
(message "Loading %s..." load-file-name)

(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(when (file-exists-p custom-file)
  (load custom-file))

(setq use-package-minimum-reported-time 0.000001)
(setq use-package-verbose t)

(use-package forge
  :after magit
  :init
  (message "++ in init.el in (use-package forge ...)")
  (message "++ setting forge-add-default-bindings to nil")
  (setq forge-add-default-sections nil)
  (setq forge-add-default-bindings nil))

(use-package magit)

(message "Loading %s...done" load-file-name)
```

then I get:

```
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/early-init.el...done
== autoloading (defvar forge-add-default-bindings...
== setting unbound forge-add-default-bindings to t
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/init.el...
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/custom.el (source)...done
Loading package magit...
-- while evaling autoloaded (with-eval-after-load ’magit-mode
-- forge-add-default-bindings is t
slightly BAD add default bindings to magit-mode-map
++ in init.el in (use-package forge ...)
++ setting forge-add-default-bindings to nil
Loading package forge...
== proper loading (defvar forge-add-default-bindings...
-- while loading forge.el: forge-add-default-bindings is nil
-- while evaling autoloaded (with-eval-after-load ’magit-mode
-- forge-add-default-bindings is nil
Loading package forge...done (0.489s)
Loading package magit...done (1.785s)
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/init.el...done
Turning on magit-auto-revert-mode...done
```

I.e. `magit-mode-map` is modified but the change that is problematic for `evil-collection` still does not occur.

**Comment by tarsius** [comment]
I then used this `init.el`:

``` emacs-lisp
(message "Loading %s..." load-file-name)

(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(when (file-exists-p custom-file)
  (load custom-file))
  
(setq evil-want-integration t)
(setq evil-want-keybinding nil)
(setq evil-collection-mode-list '(magit forge))
(require 'evil)
(when (require 'evil-collection nil t)
  (evil-collection-init))
  
(setq use-package-minimum-reported-time 0.000001)
(setq use-package-verbose t)

(use-package forge
  :after magit)

(use-package magit)

(message "Loading %s...done" load-file-name)
```

expecting `evil-mode` to kick in.  I have never used `evil`, and it seems my expectations is wrong. I am not posting the output for now but needless to the default bindings are added now.

It seems at this point you need to tell how to enable `evil` for `magit` and `forge`. In order to force `evil-collection-magit` and `evil-collection-forge` to be loaded I had to use this:

``` emacs-lisp
(setq evil-want-integration t)
(setq evil-want-keybinding nil)
(setq evil-collection-mode-list '(magit forge))
(require 'evil)
(when (require 'evil-collection nil t)
  (evil-collection-init))
;; Note that this is a testing configuration and I started emacs with:
;; HOME=/home/jonas/emacs-profiles/melpa emacs&
(add-to-list
 'load-path
 "/home/jonas/emacs-profiles/melpa/.emacs.d/elpa/evil-collection-20211120.408/modes/magit/")
(require 'evil-collection-magit)
(add-to-list
 'load-path
 "/home/jonas/emacs-profiles/melpa/.emacs.d/elpa/evil-collection-20211120.408/modes/forge/")
(require 'evil-collection-forge)
```

Please show me a modification of my demo `init.el` that succeeds in enabling `evil` for `forge` and `magit`.

**Comment by tarsius** [comment]
Now lets try `evil`:

``` emacs-lisp
(message "Loading %s..." load-file-name)

(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(when (file-exists-p custom-file)
  (load custom-file))

(setq evil-want-integration t)
(setq evil-want-keybinding nil)
(setq evil-collection-mode-list '(magit forge))
(require 'evil)
(when (require 'evil-collection nil t)
  (evil-collection-init))

(setq use-package-minimum-reported-time 0.000001)
(setq use-package-verbose t)

(use-package forge
  :after magit)

(use-package magit)

(message "Loading %s...done" load-file-name)
```

results in:

```
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/early-init.el...done
== autoloading (defvar forge-add-default-bindings...
== setting unbound forge-add-default-bindings to t
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/init.el...
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/custom.el (source)...done
Loading package magit...
-- while evaling autoloaded (with-eval-after-load ’magit-mode
-- forge-add-default-bindings is t
slightly BAD add default bindings to magit-mode-map
--evil-- loading evil-collection-magit.el
Loading package forge...
== proper loading (defvar forge-add-default-bindings...
-- while loading forge.el: forge-add-default-bindings is t
-- while evaling autoloaded (with-eval-after-load ’magit-mode
-- forge-add-default-bindings is t
slightly BAD add default bindings to magit-mode-map
BAD add default bindings to elsewhere
Cannot insert ("N" "Forge" forge-dispatch) into magit-dispatch; o not found
Loading package forge...done (0.439s)
Loading package magit...done (1.478s)
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/init.el...done
Turning on magit-auto-revert-mode...done
```

After modifying `evil-collection-magit` to contain:

``` emacs-lisp
;;; Code:

(message "--evil-- loading evil-collection-magit.el")
(require 'evil-collection)
(setq forge-add-default-sections nil)
(setq forge-add-default-bindings nil)
(require 'magit nil t)
```

we get:

``` emacs-lisp
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/early-init.el...done
== autoloading (defvar forge-add-default-bindings...
== setting unbound forge-add-default-bindings to t
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/init.el...
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/custom.el (source)...done
Loading package magit...
-- while evaling autoloaded (with-eval-after-load ’magit-mode
-- forge-add-default-bindings is t
slightly BAD add default bindings to magit-mode-map
--evil-- loading evil-collection-magit.el
Loading package forge...
== proper loading (defvar forge-add-default-bindings...
-- while loading forge.el: forge-add-default-bindings is nil
-- while evaling autoloaded (with-eval-after-load ’magit-mode
-- forge-add-default-bindings is nil
Loading package forge...done (0.426s)
Loading package magit...done (1.462s)
Loading /home/jonas/emacs-profiles/melpa/.emacs.d/init.el...done
Turning on magit-auto-revert-mode...done
```

So it works with `evil` to. You just have to do it before `magit` is loaded.

Are the changes to `magit-mode-map` a problem for `evil`?
